### ISSUE TITLE
android.media.MediaCodec$CodecException: error 0xfffffff4 / 0xffffffe0

### ISSUE THREAD

**Issue by VishnuprasadQp** [androidx/media#1789]
### Version

Media3 main branch

### More version details

Version 1.4.1

### Devices that reproduce the issue

**Nothing 2a**  running on Android 14 
Model number: Nothing A142

### Reproducible in the demo app?

Yes

### Reproduction steps

1. The issue is reproducible when playing multiple DRM content one after the other on different OTT applications. For example, I played DRM content in the Media3 Demo app, moved that app to the background, and then started playing content on Netflix and continued to play such DRM content in other OTT applications. On returning to the Demo app, I can see the error.
2. The issue is reproducible on L1 devices with MediaTek processors.
3. Once the mobile device reaches its memory threshold, the issue can occur while playing any DRM content in the Media3 Demo application.
4. The issue is easier to reproduce when PIP mode is enabled on other OTT applications.

### Expected result

The media should play successfully.

### Actual result

Playback fails with the exception.
**
Caused by: androidx.media3.exoplayer.mediacodec.MediaCodecRenderer$DecoderInitializationException: Decoder init failed: c2.mtk.avc.decoder.secure, Format(1, null, null, video/avc, avc1.640033, 26984312, null, [3840, 1714, 24.0, ColorInfo(Unset color space, Unset color range, Unset color transfer, false, 8bit Luma, 8bit Chroma)], [-1, -1])
Caused by: android.media.MediaCodec$CodecException: Failed to initialize c2.mtk.avc.decoder.secure, error 0xfffffff4 (NO_MEMORY)**
![Screenshot_20241004-125534](https://github.com/user-attachments/assets/b2fcfc2d-5ca5-44da-a1d6-3391b6ba5a7d)

### Media

I will send the additional artefacts with the error logs, etc via mail.

### Bug Report

- [x] You will email the zip file produced by `adb bugreport` to android-media-github@google.com after filing this issue.

**Comment by VishnuprasadQp** [comment]
Some applications recommend their users to restart the device. For example: JioCinema
![Screenshot_20241008-124319](https://github.com/user-attachments/assets/cb9f9de8-da38-4793-a587-bd1118f7f631)
Is it a good approach?

**Comment by VishnuprasadQp** [comment]
I have shared the bug report via mail. Please let me know if you require more details about the issue.

**Comment by rrfrias** [comment]
Do we have your permission to share this bugreport with MediaTek

Internal tracker: [376329580](http://b/376329580)

**Comment by VishnuprasadQp** [comment]
Hi Androidx/Media Team,

     Thank you for reaching out. Yes, you can share this bug report with
MediaTek.


With regards
Vishnuprasad G

**Comment by rrfrias** [comment]
Thanks. We are trying to get the right people in MediaTek to take a look at this.

**Comment by VishnuprasadQp** [comment]
Yes, if I play PIP content and switch to another application that requires
the decoder, the issue occurs.

**Comment by rrfrias** [comment]
Unfortunately device limitations is causing codec initialization to fail. We are trying to see if there is a better API to query than [MediaCodecInfo.CodecCapabilities#getMaxSupportedInstances](https://developer.android.com/reference/android/media/MediaCodecInfo.CodecCapabilities#getMaxSupportedInstances()). You might want to query this value and then err on the side of caution when initialization multiple codecs.

**Comment by VishnuprasadQp** [comment]
I already added the API, and the maximum supported instance is 1 but how can we know the codec is in use and no codecs remain to decrypt our DRM content?

**Comment by rrfrias** [comment]
We are waiting on an internal response. Ian, do you know if there is a way to check if another app is holding a decode instance? https://github.com/androidx/media/issues/1789#issuecomment-2474288973

**Comment by rrfrias** [comment]
Android is aware that given the current APIs, it is difficult for apps to know whether resources are available. There also is a service that will reclaim codec from background applications but somehow that is not triggering in your particular scenario. 

They are considering alternatives to address this, but do not have anything that we can share at the moment. You might have to catch the exception and determine that it is a case of insufficient resources. Since that is the best we can suggest at this point, i am going to close this tracker.

**Comment by MarimuthuP** [comment]
I used the following changes to resolve the above issue which allows ExoPlayer to use software decoding if hardware decoding fails,

```kotlin
exoPlayer = ExoPlayer.Builder(context)
                .setRenderersFactory(buildRenderersFactory())
                .setTrackSelector(trackSelector)
                .build()


private fun buildRenderersFactory(): RenderersFactory {
        val extensionRendererMode =
            DefaultRenderersFactory.EXTENSION_RENDERER_MODE_OFF
        return DefaultRenderersFactory(context).setExtensionRendererMode(extensionRendererMode)
            .setEnableDecoderFallback(true)
    }
